### PR TITLE
Use New Router

### PR DESCRIPTION
**Pull request by driusan**
This updates LORIS to use the PSR15 RequestHandler/Middleware based router that was implemented in PRs #3477, #3480, and #3513.

Most of the mod_rewrite rules, main.php, and NDB_Caller are replaced by a new index.php, which generates a PSR7 ServerRequestInterface and delegates to the module's Module handler.

AjaxHelper.php remains, but is deprecated, as existing scripts will need to be updated to be proper endpoints in the module's router. (Once this is done, the separate router.php for PHP's built in web server can also be removed, and mod_rewrite dependency for LORIS can be removed.)

main.php and NDB_Caller also remain for now in order to load instruments. Once an instrument module is implemented to handle the instruments routers, they can be removed.

Note that instrument URLs also change from $LORIS/instrumentname to $LORIS/instruments/instrumentname as part of this PR.

### CODE REVIEW

File: docs/config/apache2-site [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 	php_value include_path .:/usr/share/php:%LORISROOT%/project/libraries:%LORISROOT%/php/libraries
 
-	DirectoryIndex main.php index.html
+	#DirectoryIndex main.php index.html
```
Comment (xlecours) [inline docs/config/apache2-site]: is the default index.php?

File: htdocs/main.php [hunk 2]
```diff
@@ -224,7 +224,7 @@ function tplFromRequest($param)
 } finally {
     // Set dependencies if they are not set
     if (!isset($tpl_data['jsfiles']) || !isset($tpl_data['cssfiles'])) {
-        $page = (new NDB_Page(new Module(""), '', '', '', ''));
+        $page = (new NDB_Page(new Module("", ""), '', '', '', ''));
```
Comment (xlecours) [inline htdocs/main.php]: quotes uniformity.

File: modules/instrument_list/php/instrument_list.class.inc [hunk 3]
```diff
@@ -30,6 +33,9 @@ namespace LORIS\instrument_list;
  */
 class Instrument_List extends \NDB_Menu_Filter
 {
+    protected $Candidate;
```
Comment (xlecours) [inline modules/instrument_list/php/instrument_list.class.inc]: lowercase`c` ?

File: modules/instrument_list/php/instrument_list.class.inc [hunk 4]
```diff
@@ -30,6 +33,9 @@ namespace LORIS\instrument_list;
...
+    protected $TimePoint;
```
Comment (xlecours) [inline modules/instrument_list/php/instrument_list.class.inc]: lowercase `t` ?

File: modules/timepoint_list/php/timepoint_list.class.inc [hunk 5]
```diff
@@ -25,6 +27,7 @@ namespace LORIS\timepoint_list;
  */
 class Timepoint_List extends \NDB_Menu
 {
+    protected $CandID;
```
Comment (xlecours) [inline modules/timepoint_list/php/timepoint_list.class.inc]: lowercaswe `c` ?

File: php/libraries/Candidate.class.inc [hunk 6]
```diff
@@ -52,8 +52,9 @@ class Candidate
      *                        a candidate in the candidate table
      *
      * @return Candidate the Candidate object
+     * @static
      */
-    static function &singleton($candID)
+    static public function &singleton($candID)
```
Comment (xlecours) [inline php/libraries/Candidate.class.inc]: type hint : Candidate ?

File: php/libraries/NDB_BVL_Feedback.class.inc [hunk 7]
```diff
@@ -224,7 +224,7 @@ class NDB_BVL_Feedback
      *
      * @return array
      */
-    function getFeedbackTypes()
+    static function getFeedbackTypes()
```
Comment (xlecours) [inline php/libraries/NDB_BVL_Feedback.class.inc]: public
and `php/libraries/BVL_Feedback_Panel.class.inc` is making a static and a non static call

File: src/Router/ModuleFileRouter.php [hunk 8]
```diff
@@ -81,12 +83,17 @@ public function handle(ServerRequestInterface $request) : ResponseInterface
...
                 ->withBody(new \Zend\Diactoros\Stream($fullpath));
+            if ($this->contenttype != "") {
+                $resp = $resp->withHeader("Content-Type", $this->contenttype);
+            }
+            return $resp;
         }
         return (new \Zend\Diactoros\Response())
             ->withStatus(404)
-            ->withBody(new StringStream($fullpath . ": File not found"));
+            ->withBody(new StringStream($fullpath . ": File not found"))
```
Comment (xlecours) [inline src/Router/ModuleFileRouter.php]: To use the template, should be :
`return (new \LORIS\Http\Error($request, 404, $fullpath . ": File not found"));`

File: php/libraries/Candidate.class.inc (hunk 6 above)
Comment (driusan) [reply php/libraries/Candidate.class.inc]: This was just from me trying to debug something and probably shouldn't be in this PR at all.

Should I remove it? It's still a small step forward if I leave it.

Comment (xlecours) [reply php/libraries/Candidate.class.inc]: No, leave it there. There is no harm in adding public. This function, if things goes as planned, should be removed in the next 10 . years :)

File: docs/config/apache2-site (hunk 1 above)
Comment (driusan) [reply docs/config/apache2-site]: The default is "index.html", but the last rewrite rule of "RewriteRule ^(.*)$" catches everything.

**Comment by PapillonMcGill** [comment]
tried to execute but get the following error
PHP Fatal error:  Uncaught ConfigurationException: Config setting issue_tracker_url does not exist in database
I ensure that all SQL patch from 19 and 19.1 were installed and there is no SQL files in this or references PRs

**Comment by driusan** [comment]
@PapillonMcGill ? Did you make that comment on the right PR?

**Comment by driusan** [comment]
This is going to the major branch.. you're missing SQL/New_patches/2017-04-24-UseOurIssueTrackerPl0x.sql, which is unrelated to this PR

**Comment by PapillonMcGill** [comment]
@driusan Effectively, working now! will proceed with testing

**Comment by driusan** [comment]
@PapillonMcGill Thanks for catching the data team helper issue, I'll have to update those links (the URL for instruments is now /instruments/testname, I missed that module..) and I'll have to think about how to handle reliability, since it's in a weird semi-deprecated state right now..

File: src/Middleware/UserPageDecorationMiddleware.php [hunk 9]
```diff
@@ -79,14 +80,15 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
...
             $tpl_data['feedback_panel'] = $page->getFeedbackPanel(
-                $get['candID'],
+                $candID,
                 $get['sessionID'] ?? null
             );
 
-            $tpl_data['bvl_feedback'] = NDB_BVL_Feedback::bvlFeedbackPossible(
-                $this->page
+            $tpl_data['bvl_feedback'] = \NDB_BVL_Feedback::bvlFeedbackPossible(
+                $page
```
Comment (xlecours) [inline src/Middleware/UserPageDecorationMiddleware.php]: should be `$this->PageName`
(will solve `1) InstrumentListTestIntegrationTest::testPageUIs`)

File: modules/timepoint_list/php/timepoint_list.class.inc [hunk 10]
```diff
@@ -66,6 +71,32 @@ class Timepoint_List extends \NDB_Menu
...
+     * @param ServerRequestInterface  $request The incoming PSR7 request
+     * @param RequestHandlerInterface $handler The PSR15 handler.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    public function process(
+        ServerRequestInterface $request,
+        RequestHandlerInterface $handler
+    ) : ResponseInterface {
+        $attribute = $request->getAttribute("CandID");
```
Comment (xlecours) [inline modules/timepoint_list/php/timepoint_list.class.inc]: `CandID` --> `candID` 
(will solve `2) TimepointListIntegrationTest::testTimepointFilteringRules`)
Why use capital letters at all... just saying

**Comment by driusan** [comment]
@PapillonMcGill I think I've addressed the issues with the data team helper module, but I'm not familiar enough with the module and the test plan doesn't have enough detail to know if it's working properly (but the links are definitely working better than before.)

@xlecours @ridz1208 This is now passing Travis (Finally!), so it can be reviewed.

**Comment by PapillonMcGill** [comment]
@driusan Datahelper now report that "This page (pagename) is under construction"
while loading the reliability page cause a hole bunch of "undefined offset" and 3 php warning of:
` Cannot modify header information - headers already sent by (output started at /var/www/Loris/modules/reliability/php/reliability.class.inc:735) in /var/www/Loris/htdocs/index.php on line 48, referer: https://*instance*.loris.ca/reliability/`

**Comment by driusan** [comment]
@PapillonMcGill Neither the test plan nor the code for the `data_team_helper` module make any reference to reliability. Where are you talking about?

**Comment by PapillonMcGill** [comment]
@driusan 
For reliability, path is: clinical -> reliability.
on the page itself, the data table is empty with an "error loading data" box, the warning are from apache log.

**Comment by driusan** [comment]
@PapillonMcGill I don't think it's worth spending a lot of time on reliability, because it's a one project module that just can't be removed from LORIS because it can't be fully converted to a standalone module (yet, due to dependencies on NDB_Caller), but I managed to get one reliability instrument to load with the latest commit, it just falls back on main.php (like instruments for now) until it's properly converted.

File: php/libraries/NDB_BVL_Feedback.class.inc [hunk 11]
```diff
@@ -1065,25 +1065,22 @@ class NDB_BVL_Feedback
...
-                    || ($test_name === "instrument_list")
-                ) {
+        if ($user->hasPermission('bvl_feedback')) {
+            if (($test_name === "timepoint_list")
+                || ($test_name === "instrument_list")
+            ) {
+                return true;
+            } else {
+                $Factory       = NDB_Factory::singleton();
+                $DB            = $Factory->Database();
```
Comment (xlecours) [inline php/libraries/NDB_BVL_Feedback.class.inc]: the database function is lowercase

**Comment by xlecours** [comment]
@driusan, I will wait for the Travis fixes to get in.

**Pull request merged**